Support record_stream in dispatch mode

lw opened the pull request:
Summary:
Issuing a `t.record_stream(s)` call while a `TorchDispatchMode` is active was throwing because PyTorch was unable to convert a c10::Stream back to a Python object. It's now fixed.

Fixes https://github.com/pytorch/pytorch/issues/94403

Test Plan: Added a unit test

Differential Revision: D45117566

albanD commented on torch/csrc/Stream.cpp:
```diff
@@ -43,6 +43,22 @@ static PyObject* THPStream_pynew(
+  self->stream_id = stream.id();
+  self->device_index = stream.device_index();
+  self->device_type = static_cast<int64_t>(stream.device_type());
+  return (PyObject*)ptr.release();
```
This is already a `PyObject*` no need for the cast

albanD commented on torch/csrc/Stream.cpp:
```diff
@@ -43,6 +43,22 @@ static PyObject* THPStream_pynew(
   END_HANDLE_TH_ERRORS
 }
 
+PyObject* THPStream_New(const c10::Stream& stream) {
```
The `new` name is usually reserved for the tp_new slow function. We usually use `wrap()` for the converter from c++ object to python object. Can you rename this?

albanD reviewed (commented):
Sounds pretty good!

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -13686,6 +13686,13 @@
   tags: view_copy
   autogen: _test_autograd_multiple_dispatch_view_copy.out
 
+# Note: this function is only for testing in TestPythonDispatch.test_return_stream
```
Ho you don't need a brand new native function for this, you can define it directly from python for testing:
```python
import torch
from torch.library import Library

l = Library("aten", "DEF")
l.define("_test_return_stream() -> Stream")

# If you need an implementation, you can register it as l.imp()
# You can then call this native op via torch.ops.aten._test_return_stream just like this one.

```

albanD reviewed (commented):
Look good, just one change on how to define the test function.
Sorry I should have been clearer about that in my comment above.

albanD reviewed (approved):
Much simpler!
Thanks for taking the time to update the test!

lw commented:
@pytorchbot merge

lw committed:
Support record_stream in dispatch mode (#99529)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/99529

Issuing a `t.record_stream(s)` call while a `TorchDispatchMode` is active was throwing because PyTorch was unable to convert a c10::Stream back to a Python object. It's now fixed.

Fixes https://github.com/pytorch/pytorch/issues/94403

Test Plan: Added a unit test

Differential Revision: D45117566

fbshipit-source-id: 240270fd4f415c1ca0c856d8ccd19b527c4ce6bd
